### PR TITLE
Implement search functionality

### LINKED ISSUE #839: Implement search functionality
Remove the auto-suggestion thing. After a user types something and clicks search. Perform string matching to see if it fits into what and route to that page.

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed #839

### CODE REVIEW

File: scan/__tests__/Route_test.re [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  test("test block route", () =>
+    expect("123" |> search |> toString) |> toEqual("/block/123")
+  );
+  test("test block route", () =>
+    expect(
+      "22638794cb5f306ef929b90c58b27d26cb35a77ca5c5c624cf2025a98528c323" |> search |> toString,
+    )
+    |> toEqual("/tx/22638794cb5f306ef929b90c58b27d26cb35a77ca5c5c624cf2025a98528c323")
+  );
+  test("test block prefix is B", () =>
```
Comment (sorawit) [inline scan/__tests__/Route_test.re:25]: Let's have this compared route directly if possible? We should not have tests that rely on `toString` unless necesssary.

File: scan/src/utils/Route.re [hunk 2]
```diff
@@ -123,3 +123,56 @@ let toString =
   | NotFound => "/";
 
 let redirect = (route: t) => ReasonReactRouter.push(route |> toString);
+
+let search = (str: string) => {
+  let len = str |> String.length;
+  let capStr = str |> String.capitalize_ascii;
+  // let test = str |> Js_Re.startsWith("O");
+  let blockID =
+    capStr |> Js.String.startsWith("B")
```
Comment (sorawit) [inline scan/src/utils/Route.re]: Can we do

```
let search = (str) => {
    ...
    if (capStr |> Js.String.startswith("B")) {
        let%Opt blockID = String.sub(_, 1, len - 1) |> int_of_string_opt;
        Some(BlockIndexPage(blockID))
    } else if (...) {
        ...
    } |> Belt.Option.getWithDefault(_, NotFound)
}
```

**Review by sorawit — CHANGES_REQUESTED** [review]

File: scan/src/utils/Route.re [hunk 3]
```diff
@@ -123,3 +123,37 @@ let toString =
...
+      } else if (capStr |> Js.String.startsWith("D")) {
+        let%Opt dataSourceID = str |> String.sub(_, 1, len - 1) |> int_of_string_opt;
+        Some(DataSourceIndexPage(dataSourceID, DataSourceExecute));
+      } else if (capStr |> Js.String.startsWith("R")) {
+        let%Opt requestID = str |> String.sub(_, 1, len - 1) |> int_of_string_opt;
+        Some(RequestIndexPage(requestID, RequestReportStatus));
+      } else if (capStr |> Js.String.startsWith("O")) {
+        let%Opt oracleScriptID = str |> String.sub(_, 1, len - 1) |> int_of_string_opt;
+        Some(ScriptIndexPage(oracleScriptID, ScriptTransactions));
+      } else if (len == 64) {
```
Comment (sorawit) [inline scan/src/utils/Route.re]: Let's also allow the case when string is `0x` prefixed.

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — CHANGES_REQUESTED** [review]
Two things.
- We should also redirect to the search page if user press "Enter"
- We should clear the search box after we redirect

File: scan/src/components/SearchBar.re [hunk 4]
```diff
@@ -195,16 +195,30 @@ let make = () => {
         | _ => ()
         }
       }
+      onKeyPress={event =>
```
Comment (sorawit) [inline scan/src/components/SearchBar.re]: Let's do this logic in `onKeyDown`

**Review by sorawit — CHANGES_REQUESTED** [review]

File: scan/src/utils/Route.re [hunk 5]
```diff
@@ -123,3 +123,37 @@ let toString =
...
+      } else if (capStr |> Js.String.startsWith("D")) {
+        let%Opt dataSourceID = str |> String.sub(_, 1, len - 1) |> int_of_string_opt;
+        Some(DataSourceIndexPage(dataSourceID, DataSourceExecute));
+      } else if (capStr |> Js.String.startsWith("R")) {
+        let%Opt requestID = str |> String.sub(_, 1, len - 1) |> int_of_string_opt;
+        Some(RequestIndexPage(requestID, RequestReportStatus));
+      } else if (capStr |> Js.String.startsWith("O")) {
+        let%Opt oracleScriptID = str |> String.sub(_, 1, len - 1) |> int_of_string_opt;
+        Some(ScriptIndexPage(oracleScriptID, ScriptTransactions));
+      } else if (len == 64 || str |> Js.String.startsWith("0x") && len == 66) {
```
Comment (sorawit) [inline scan/src/utils/Route.re:153]: Let's add paren here for clarity. `xxx && (xxx || xxx)`

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by sorawit — DISMISSED** [review]

File: scan/src/utils/Route.re [hunk 6]
```diff
@@ -123,3 +123,37 @@ let toString =
...
+let search = (str: string) => {
+  let len = str |> String.length;
+  let capStr = str |> String.capitalize_ascii;
+
+  (
+    switch (str |> int_of_string_opt) {
+    | Some(blockID) => Some(BlockIndexPage(blockID))
+    | None =>
+      if (str |> Js.String.startsWith("bandvaloper")) {
+        Some(ValidatorIndexPage(str |> Address.fromBech32, Delegators));
```
Comment (taobun) [inline scan/src/utils/Route.re]: Go to proposed block?

**Review by taobun — CHANGES_REQUESTED** [review]

File: scan/src/pages/ScriptIndexPage.re [hunk 7]
```diff
@@ -120,7 +120,9 @@ let make = (~codeHash, ~hashtag: Route.script_tab_t) => {
     // />
     <VSpacing size=Spacing.xl />
     <Row>
-      <Col> <InfoHL info={InfoHL.Hash(codeHash, Colors.purple3)} header="SCRIPT HASH" /> </Col>
+      <Col>
+        <InfoHL info={InfoHL.Text(oracleScriptID |> string_of_int)} header="SCRIPT HASH" />
```
Comment (taobun) [inline scan/src/pages/ScriptIndexPage.re]: It will be changed in other pr

File: scan/__tests__/Route_test.re [hunk 8]
```diff
@@ -0,0 +1,70 @@
+open Jest;
+open Route;
+open Expect;
+
+describe("Expect Search Functionality to work correctly", () => {
+  test("test block route", () =>
+    expect("123" |> search) |> toEqual(BlockIndexPage(123))
+  );
+  test("test block route", () =>
```
Comment (taobun) [inline scan/__tests__/Route_test.re]: transaction

File: scan/__tests__/Route_test.re [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    expect("22638794cb5f306ef929b90c58b27d26cb35a77ca5c5c624cf2025a98528c323" |> search)
+    |> toEqual(
+         TxIndexPage(
+           "22638794cb5f306ef929b90c58b27d26cb35a77ca5c5c624cf2025a98528c323" |> Hash.fromHex,
+         ),
+       )
+  );
+  test("test block route prefix is 0x", () =>
```
Comment (taobun) [inline scan/__tests__/Route_test.re]: transaction

File: scan/__tests__/Route_test.re [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  test("test validator route", () =>
+    expect("bandvaloper1p40yh3zkmhcv0ecqp3mcazy83sa57rgjde6wec" |> search)
+    |> toEqual(
+         ValidatorIndexPage(
+           "bandvaloper1p40yh3zkmhcv0ecqp3mcazy83sa57rgjde6wec" |> Address.fromBech32,
+           ProposedBlocks,
+         ),
+       )
+  );
+  test("test oracle script route", () =>
```
Comment (taobun) [inline scan/__tests__/Route_test.re]: account

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**